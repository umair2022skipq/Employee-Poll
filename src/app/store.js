import { configureStore } from "@reduxjs/toolkit";
import employeePollReducer from "../features/employeePoll/employeePollSlice";
import userSliceReducer from "../features/userSlice/userSlice";
import authSliceReducer from "../features/authSlice/authSlice";

export default configureStore({
  reducer: {
    poll: employeePollReducer,
    user: userSliceReducer,
    auth: authSliceReducer,
  },
});
