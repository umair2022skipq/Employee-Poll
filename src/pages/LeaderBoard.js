import { List } from "@mui/material";
import { ListItem, ListItemAvatar, ListItemText } from "@mui/material";
import { Avatar } from "@mui/material/";
import Typography from "@mui/material/Typography";
import { useSelector } from "react-redux/es/exports";
import { leaderSelector } from "../features/user/userSlice";
import { Spinner } from "./NewQuestions";
import { authSelector } from "../features/auth/authSlice";
import { userSelector } from "../features/user/userSlice";

const LeaderBoard = ({ isLoading }) => {
  const leaders = useSelector(leaderSelector);
  const auth = useSelector(authSelector);
  const user = useSelector(userSelector);

  const authUser = auth.userId ? user.users.byId[auth.userId] : null;
  console.log(authUser);

  if (isLoading) {
    return <Spinner />;
  }

  return (
    <>
      <Typography variant="h2" component="div" gutterBottom>
        Leaderboard
      </Typography>
      {leaders &&
        leaders.map((user) => (
          <List key={user.id}>
            <ListItem
              style={{
                backgroundColor: user.id === authUser.id ? "#bcffc7" : "",
              }}
            >
              <ListItemAvatar>
                <Avatar src={user.avatarURL || ""} />
              </ListItemAvatar>
              <ListItemText
                primary={`${user.name}`}
                secondary={
                  <>
                    {`Current Score: ${
                      user.questions.length + Object.keys(user.answers).length
                    }`}{" "}
                    <br />
                    {`Questions Asked: ${user.questions.length}`} <br />
                    {`Questions Answered: ${Object.keys(user.answers).length}`}
                  </>
                }
              />
            </ListItem>
          </List>
        ))}
    </>
  );
};

export default LeaderBoard;
