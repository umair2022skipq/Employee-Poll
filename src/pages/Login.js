import Avatar from "@mui/material/Avatar";
import CssBaseline from "@mui/material/CssBaseline";
import TextField from "@mui/material/TextField";
import Link from "@mui/material/Link";
import { Container, Alert } from "@mui/material";
import Box from "@mui/material/Box";
import LockOutlinedIcon from "@mui/icons-material/LockOutlined";
import Typography from "@mui/material/Typography";
import { createTheme, ThemeProvider } from "@mui/material/styles";
import { useDispatch, useSelector } from "react-redux";
import { authSelector, loginAsync } from "../features/auth/authSlice";
import LoadingButton from "@mui/lab/LoadingButton";

const Login = () => {
  function Copyright(props) {
    return (
      <Typography
        variant="body2"
        color="text.secondary"
        align="center"
        {...props}
      >
        {"Copyright Â© "}
        <Link color="inherit" href="https://erdemkeren.com/">
          Hilmi Erdem Keren
        </Link>{" "}
        {new Date().getFullYear()}
        {"."}
      </Typography>
    );
  }
  const auth = useSelector(authSelector);
  const dispatch = useDispatch();

  const handleSubmit = (event) => {
    event.preventDefault();
    const data = new FormData(event.currentTarget);
    const username = data.get("username");
    const password = data.get("password");

    // if (typeof username !== "string" || typeof password !== "string") {
    //   return;
    // }
    dispatch(loginAsync({ username, password }));
  };

  const theme = createTheme();

  return (
    <ThemeProvider theme={theme}>
      <Container component="main" maxWidth="xs">
        <CssBaseline />
        <Box
          sx={{
            marginTop: 8,
            display: "flex",
            flexDirection: "column",
            alignItems: "center",
          }}
        >
          <Avatar sx={{ m: 1, bgcolor: "secondary.main" }}>
            <LockOutlinedIcon />
          </Avatar>
          <Typography component="h1" variant="h5">
            Sign in
          </Typography>
          <Box
            component="form"
            onSubmit={handleSubmit}
            noValidate
            sx={{ mt: 1 }}
          >
            <TextField
              margin="normal"
              required
              fullWidth
              id="username"
              label="Username"
              name="username"
              autoComplete="username"
              defaultValue="sarahedo"
              autoFocus
            />
            <TextField
              margin="normal"
              required
              fullWidth
              name="password"
              label="Password"
              type="password"
              id="password"
              autoComplete="current-password"
              defaultValue="password123"
            />
            {auth.status === "failed" && (
              <Alert severity="warning">
                The authentication was not successful.
              </Alert>
            )}
            <LoadingButton
              type="submit"
              fullWidth
              variant="contained"
              sx={{ mt: 3, mb: 2 }}
              loading={auth.status === "loading"}
            >
              Sign In
            </LoadingButton>
          </Box>
        </Box>
        <Copyright sx={{ mt: 8, mb: 4 }} />
      </Container>
    </ThemeProvider>
  );
};
export default Login;
