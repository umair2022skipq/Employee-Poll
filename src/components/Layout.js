import { createTheme, ThemeProvider } from "@mui/material/styles";
import CssBaseline from "@mui/material/CssBaseline";
import Box from "@mui/material/Box";
import Toolbar from "@mui/material/Toolbar";
import List from "@mui/material/List";
import Typography from "@mui/material/Typography";
import Divider from "@mui/material/Divider";
import IconButton from "@mui/material/IconButton";
import Container from "@mui/material/Container";
import MenuIcon from "@mui/icons-material/Menu";
import ChevronLeftIcon from "@mui/icons-material/ChevronLeft";
import LogoutIcon from "@mui/icons-material/Logout";
import ListItem from "@mui/material/ListItem";
import ListItemIcon from "@mui/material/ListItemIcon";
import ListItemText from "@mui/material/ListItemText";
import Avatar from "@mui/material/Avatar";

import { AppBar, Drawer } from "../style/styles";

import { useState } from "react";
import { Routes, Route } from "react-router-dom";
import { useSelector, useDispatch } from "react-redux";

import { MainListItems } from "./ListItems";
import NewQuestions from "../pages/NewQuestions";
import LeaderBoard from "../pages/LeaderBoard";
import AddQuestion from "../pages/AddQuestion";
import QuestionDetails from "../pages/QuestionDetails";
import Copyright from "./Copyright";

import { authSelector, logout } from "../features/auth/authSlice";
import { userSelector } from "../features/user/userSlice";
import { useNavigate } from "react-router-dom";

const Layout = ({ isLoading }) => {
  const mdTheme = createTheme();

  const auth = useSelector(authSelector);
  const user = useSelector(userSelector);

  const dispatch = useDispatch();

  const navigate = useNavigate();

  const authUser = auth.userId ? user.users.byId[auth.userId] : null;

  const [open, setOpen] = useState(true);
  const toggleDrawer = () => {
    setOpen(!open);
  };

  const logoutHandler = () => {
    dispatch(logout());
    navigate("/");
  };

  if (!authUser) {
    return (
      <Typography
        component="h1"
        variant="h6"
        color="inherit"
        noWrap
        sx={{ flexGrow: 1, ml: 1 }}
      >
        {" "}
        User does not exist{" "}
      </Typography>
    );
  }

  return (
    <ThemeProvider theme={mdTheme}>
      <Box sx={{ display: "flex" }}>
        <CssBaseline />
        <AppBar position="absolute" open={open}>
          <Toolbar
            sx={{
              pr: "24px", // keep right padding when drawer closed
            }}
          >
            <IconButton
              edge="start"
              color="inherit"
              aria-label="open drawer"
              onClick={toggleDrawer}
              sx={{
                marginRight: "36px",
                ...(open && { display: "none" }),
              }}
            >
              <MenuIcon />
            </IconButton>
            <Typography
              component="h1"
              variant="h6"
              color="inherit"
              noWrap
              sx={{ flexGrow: 1 }}
            >
              Employes Poll
            </Typography>
            <Box sx={{ flexGrow: 0, display: "flex" }}>
              <div>
                <IconButton
                  size="large"
                  aria-label="account of current user"
                  aria-controls="menu-appbar"
                  aria-haspopup="true"
                  color="inherit"
                >
                  <Avatar
                    alt={authUser.name || undefined}
                    src={authUser.avatarURL || undefined}
                  />

                  <Typography
                    component="h1"
                    variant="h6"
                    color="inherit"
                    noWrap
                    sx={{ flexGrow: 1, ml: 1 }}
                  >
                    {authUser.name}
                  </Typography>
                </IconButton>
              </div>
            </Box>
          </Toolbar>
        </AppBar>
        <Drawer variant="permanent" open={open}>
          <Toolbar
            sx={{
              display: "flex",
              alignItems: "center",
              justifyContent: "flex-end",
              px: [1],
            }}
          >
            <IconButton onClick={toggleDrawer}>
              <ChevronLeftIcon />
            </IconButton>
          </Toolbar>
          <Divider />
          <List component="nav">
            <MainListItems />
          </List>
          <Divider sx={{ my: 1 }} />
          <List>
            <ListItem button onClick={logoutHandler}>
              <ListItemIcon>
                <LogoutIcon />
              </ListItemIcon>
              <ListItemText
                primary={
                  <Typography variant="body1" display="block" gutterBottom>
                    Logout <b>{authUser.name}</b>
                  </Typography>
                }
              />
            </ListItem>
          </List>
        </Drawer>
        <Box
          component="main"
          sx={{
            backgroundColor: (theme) =>
              theme.palette.mode === "light"
                ? theme.palette.grey[100]
                : theme.palette.grey[900],
            flexGrow: 1,
            height: "100vh",
            overflow: "auto",
          }}
        >
          <Toolbar />
          <Container maxWidth="lg" sx={{ mt: 4, mb: 4 }}>
            <Routes>
              <Route
                path="/"
                element={<NewQuestions isLoading={isLoading} />}
              />
              <Route
                path="/answered"
                element={<NewQuestions isLoading={isLoading} />}
              />
              <Route path="/add" element={<AddQuestion />} />
              <Route path="/leaderboard" element={<LeaderBoard />} />
              <Route
                path="/questions/:question_id"
                element={<QuestionDetails isLoading={isLoading} />}
              />
            </Routes>
          </Container>
          <Copyright sx={{ pt: 4 }} />
        </Box>
      </Box>
    </ThemeProvider>
  );
};

export default Layout;
