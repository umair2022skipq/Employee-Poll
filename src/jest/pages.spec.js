import { render, fireEvent, queryByAttribute } from "@testing-library/react";
import Login from "../pages/login";
import Layout from "../components/layout";
import { Provider } from "react-redux";
import store from "../app/store";
import LeaderBoard from "../pages/LeaderBoard";
import { _getUsers, _getQuestions } from "../data/_DATA";
import { BrowserRouter as Router } from "react-router-dom";

describe("Testing snapshots pages", () => {
  it("should render login", function () {
    const { container } = render(
      <Provider store={store}>
        <Login />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it("should render leaderboard", function () {
    const { container } = render(
      <Provider store={store}>
        <LeaderBoard />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it("should render layout", function () {
    const { container } = render(
      <Router>
        <Provider store={store}>
          <Layout />
        </Provider>
      </Router>
    );

    expect(container).toMatchSnapshot();
  });

  it("should login", function () {
    const { container } = render(
      <Provider store={store}>
        <Login />
      </Provider>
    );

    const getById = queryByAttribute.bind(null, "id");

    const username = getById(container, "username");
    const password = getById(container, "password");

    fireEvent.change(username, { target: { value: "sarahedo" } });
    fireEvent.change(password, { target: { value: "password123" } });

    const button = getById(container, ":r5:");
    fireEvent.click(button);
  });
});

describe("Testing users and questions", () => {
  it("fetch users", async () => {
    await expect(_getUsers()).not.toBeUndefined();
  });

  it("fetch questions", async () => {
    await expect(_getQuestions()).not.toBeUndefined();
  });
});
