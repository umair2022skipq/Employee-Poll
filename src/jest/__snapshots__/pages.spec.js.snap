// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing snapshots pages should render 1`] = `
<div>
  <main
    class="MuiContainer-root MuiContainer-maxWidthXs css-nzinn5-MuiContainer-root"
  >
    <div
      class="MuiBox-root css-binzgt"
    >
      <div
        class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1999axt-MuiAvatar-root"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="LockOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zM9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9V6zm9 14H6V10h12v10zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2z"
          />
        </svg>
      </div>
      <h1
        class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
      >
        Sign in
      </h1>
      <form
        class="MuiBox-root css-164r41r"
        novalidate=""
      >
        <div
          class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-focused Mui-required css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="username"
            id="username-label"
          >
            Username
            <span
              aria-hidden="true"
              class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              autocomplete="mtsamis"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="username"
              name="username"
              required=""
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1in441m"
              >
                <span>
                  Username
                   
                  *
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="password"
            id="password-label"
          >
            Password
            <span
              aria-hidden="true"
              class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              autocomplete="current-password"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="password"
              name="password"
              required=""
              type="password"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1ftyaf0"
              >
                <span>
                  Password
                   
                  *
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button
          class="MuiLoadingButton-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root css-1qudo6r-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
          id=":r2:"
          tabindex="0"
          type="submit"
        >
          Sign In
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </form>
    </div>
    <p
      class="MuiTypography-root MuiTypography-body2 MuiTypography-alignCenter css-ijul54-MuiTypography-root"
    >
      Copyright © 
      <a
        class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1cghntr-MuiTypography-root-MuiLink-root"
        href="/"
      >
        Employee Poll
      </a>
       
      2022
      .
    </p>
  </main>
</div>
`;

exports[`Testing snapshots pages should render 2`] = `
<div>
  <div
    class="MuiTypography-root MuiTypography-h2 MuiTypography-gutterBottom css-i1hrn3-MuiTypography-root"
  >
    Leaderboard
  </div>
</div>
`;

exports[`Testing snapshots pages should render 3`] = `
<div>
  <p
    class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
  >
     User does not exist 
  </p>
</div>
`;

exports[`Testing snapshots pages should render layout 1`] = `
<div>
  <p
    class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
  >
     User does not exist 
  </p>
</div>
`;

exports[`Testing snapshots pages should render leaderboard 1`] = `
<div>
  <div
    class="MuiTypography-root MuiTypography-h2 MuiTypography-gutterBottom css-i1hrn3-MuiTypography-root"
  >
    Leaderboard
  </div>
</div>
`;
